.circleci/config.yml

```diff
@@ -151,7 +154,10 @@ jobs:
             mv .coverage .coverage.imitation
             coverage combine  # rewrite paths from virtualenv to src/
       - codecov/upload
-
+      - save_cache:
```

Comment:
It appears that CircleCI cache is immutable: https://circleci.com/docs/2.0/caching/#writing-to-the-cache-in-workflows So IIUC as written this will only save the cache the first time this runs, and it will never be invalidated again, until we bump `v1` to `v2` etc. Is this intended?

One option would be to checksum on `setup.py` like we do for dependencies. This will result in invalidating it sometimes when it's not necessary, though, probably only worthwhile if it's quick to regenerate.

Reply:
That is true. I deliberately decided to just add the version number to keep it as simple as possible. However, I did not consider the problem, that we might silently retrain the experts for each test run, which would be wasteful.

I added the `setup.py` just as you recommended and also added a more generic key to the `restore_cache` option. This ensures, that whenever we change `setup.py`, we either have a cache miss (never trained anything previously) or just fetch the previous pytest cache. Either this cached experts are still properly loadable (then they are just copied to the new circleci cache) or they are retrained (and then placed in the new circleci cache). This is a nice way to prevent retraining when the `setup.py` changes are not relevant for the policy.

Now we still need to hope that changes outside of `setup.py` will cause or policies to not be loadable again.

Reply:
I updated the comments and also linked the docs just in case.